tests/unit/test_model_setup.py

```diff
@@ -0,0 +1,918 @@
+    bc_val_prim_1 = np.ravel(bc_val_prim_1, "F")
+
+    # Case 2
+    bc_val_prim_2 = np.array([0 * ones, 0 * ones])
+    bc_val_prim_2[0, east] = 0.01
+    bc_val_prim_2[1, east] = 0
+    bc_val_prim_2[0, west] = -0.01
+    bc_val_prim_2[1, west] = 0
+    bc_val_prim_2[0, south] = 0
+    bc_val_prim_2[1, south] = 0.1
+    bc_val_prim_2[0, north] = 0
+    bc_val_prim_2[1, north] = -0.1
+    bc_val_prim_2 = np.ravel(bc_val_prim_2, "F")
+
+    bc_mech_values_primary = [bc_val_prim_0, bc_val_prim_1, bc_val_prim_2]
+
+    # -----> Reference scalar
+    reference_scalar = [np.ones(2), 2.0 * np.ones(2), 100.0 * np.ones(2)]
+
+    # -----> Populating list of model parameters
+    model_params = []
+    before_runs = [True, False]
+    for before_run in before_runs:
+        for case in range(3):
+            model_param = {
+                "use_ad": True,
+                "case": case,
+                "before_run": before_run,
+                "mdg": mdg,
+                "box": box,
+                "time_step": time_steps[case],
+                "end_time": end_times[case],
+                "bc_flow_type_primary": bc_flow_type_primary[case],
+                "bc_flow_type_secondary": bc_flow_type_secondary[case],
+                "bc_flow_values_primary": bc_flow_values_primary[case],
+                "bc_flow_values_secondary": bc_flow_values_secondary[case],
+                "source_flow_primary": source_flow_primary[case],
+                "source_flow_secondary": source_flow_secondary[case],
+                "storativity_primary": storativity_primary[case],
+                "storativity_secondary": storativity_secondary[case],
+                "aperture_primary": aperture_primary[case],
+                "aperture_secondary": aperture_secondary[case],
+                "biot_alpha_primary": biot_alpha_primary[case],
+                "biot_alpha_secondary": biot_alpha_secondary[case],
+                "permeability_primary": permeability_primary[case],
+                "permeability_secondary": permeability_secondary[case],
+                "viscosity_primary": viscosity_primary[case],
+                "viscosity_secondary": viscosity_secondary[case],
+                "vector_source_primary": vector_source_primary[case],
+                "vector_source_secondary": vector_source_secondary[case],
+                "vector_source_interface": vector_source_interface[case],
+                "reference_scalar": reference_scalar[case],
+                "bc_mech_type_primary": bc_mech_type_primary[case],
+                "bc_mech_values_primary": bc_mech_values_primary[case],
+            }
+            model_params.append(model_param)
+
+    @pytest.mark.parametrize("model_params", model_params)
+    def test_non_default_parameters_contact_mechanics_biot(self, model_params):
+        """Test model with non-default scalar sources"""
+
+        # Construct the class
+        class NonDefaultContactMechanicsBiot(pp.ContactMechanicsBiot):
```

Comment:
Wouldn't it be cleaner to define the class outside the test?

Reply:
Yes. Initially, I tried to do what you suggest, i.e., define the `NoDefaultContactMechanicsBiot` class outside the test and parametrize the models instead of the models' parameters. 

To my surprise, when I tried to do this, all models had the same parameters inside their respective data dictionaries. AFAIK, the reason for this behavior is that  when several `NoDefaultContactMechanicsBiot` instances are created outside the test (as it is done by the `pytest` parametrize decorator), the `data` dictionaries mutate along process, since they all point to the same class.